### DISCUSSION TITLE
Does readonly make your code faster?

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#103] [General]
# Does readonly make your code faster?

In this blog post we will discover whether or not the readonly modifier can make your code faster. So without further ado let's get started.

https://steven-giesel.com/blogPost/50b84029-4f1e-4e6d-8fd1-0311b0605562

**Comment by jslicer** [comment]
When I tell folks to mark their fields `readonly`, it's not for performance - it's for correctness. If the design calls for a field to only be initialized with a value, the `readonly` keyword keeps further code from tampering with that contract.

**Reply by linkdotnet** [reply]
Absolutely - this blog post was created from a conversation I had with a peer.
It reminded me that I tried to debunk that claim some years ago but couldn't find that I wrote about it. So here we are :D

In any case those low level optimization make only sense (if there would be some) in <0.1% of all cases.

**Reply by darcythomas** [reply]
I find a side effect of correctness:

1. You spend less future time on toil, fixing bugs
2. You have more confidence updating to the latest framework version (library etc)

So you now have time to:

1. Look to other ways to improve performance
2. Don't bother looking and get a performance improvement by updating to .net latest

**Comment by lassevk** [comment]
It looks like the assembly code in this post is from a Debug build, can you verify?

**Reply by linkdotnet** [reply]
Rookie mistake and yes I did.
I'll update the blog post - but to no surprise, the JIT code is still the same.

**Reply by lassevk** [reply]
I agree, the code just becomes remarkably smaller. For the lark of it, I added a volatile field, and this does introduce a difference, but I can't even remember when I had to use a volatile field the last time, I'm not writing code on that level, like zero-lock code and such.

**Reply by linkdotnet** [reply]
That makes sense as the compiler can't optimize given the `volatile` keyword.

My experience: In the last 10 years I used it in 1 project and 1 project only.

**Comment by darcythomas** [comment]
I wonder if (PGO){https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-8/#tiering-and-dynamic-pgo} would optimize `readonly` fields?

Hey if anyone is looking for a place to make an open source contribution to .Net, taking advantage of `readonly` fields, could be ripe for the picking!

**Reply by linkdotnet** [reply]
That is an interesting thought. My gut would tells me that there isn't much of a difference given that it uses actual usage statistics of methods/fields and so on. If the only difference is the `readonly` keyword, the usage is probably the same, yielding the same performance characteristics.

**Comment by lassevk** [comment]
> I wonder if (PGO){[https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-8/#tiering-and-dynamic-pgo}](https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-8/#tiering-and-dynamic-pgo%7D) would optimize `readonly` fields?
> 
> Hey if anyone is looking for a place to make an open source contribution to .Net, taking advantage of `readonly` fields, could be ripe for the picking!

I believe PGO already does this. In Stephen Toubs excellent article on [Performance Improvements in .NET 8](https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-8/), he mentions that:

> For example, the JIT knows that a method “tiering up” from tier 0 to tier 1 has already been executed, and if it’s already been executed, then any static readonly fields it accesses are now already initialized, which means the JIT can look at the values of those fields and base the tier 1 code gen on what’s actually in the field (e.g. if it’s a static readonly bool, the JIT can now treat the value of that field as if it were const bool).

**Reply by linkdotnet** [reply]
Okay I revoke my earlier statement - even though `static readonly` seems a very special case that is easier to optimize than a `readonly` instance field.

**Comment by lassevk** [comment]
> Okay I revoke my earlier statement - even though `static readonly` seems a very special case that is easier to optimize than a `readonly` instance field.

But a readonly *instance* field will still have different values across different instances, which means the JITter isn't any closer to optimize a method using that field than a non-readonly field. `static readonly` means one field, total, which is a bit simpler to optimize for.

**Reply by linkdotnet** [reply]
Even `static readonly` can be different for short-lived apps (like Azure Functions/AWS Lambdas).
Given: `static readonly Now = DateTime.Now` - I am not sure how you can handle this as "const".

**Comment by abatishchev** [comment]
Does `sealed` make your code faster though? Can you please explore?

**Reply by linkdotnet** [reply]
I got you covered: https://steven-giesel.com/blogPost/9e7fbbb6-f82a-4ca6-8f6a-b124afc30d48/why-are-sealed-classes-faster-in-c-and-should-i-seal-them

did this some time ago
